src/solvers/petsc_dm_wrapper.C

```diff
@@ -16,39 +16,355 @@
 // Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 
 #include "libmesh/libmesh_common.h"
+#include "libmesh/petsc_macro.h"
 
 #ifdef LIBMESH_HAVE_PETSC
+#if !PETSC_VERSION_LESS_THAN(3,7,3)
 
 #include "libmesh/ignore_warnings.h"
 #include <petscsf.h>
+#include <petsc/private/dmimpl.h>
 #include "libmesh/restore_warnings.h"
 
 #include "libmesh/petsc_dm_wrapper.h"
 #include "libmesh/system.h"
+#include "libmesh/mesh.h"
 #include "libmesh/mesh_base.h"
+#include "libmesh/mesh_refinement.h"
+#include "libmesh/partitioner.h"
 #include "libmesh/dof_map.h"
 #include "libmesh/elem.h"
-#include "libmesh/dof_map.h"
 
 namespace libMesh
 {
 
+  //--------------------------------------------------------------------
+  // Functions with C linkage to pass to PETSc.  PETSc will call these
+  // methods as needed.
+  //
+  // Since they must have C linkage they have no knowledge of a namespace.
+  // We give them an obscure name to avoid namespace pollution.
+  //--------------------------------------------------------------------
+  extern "C"
+  {
+
+    //! Help PETSc create a subDM given a global dm when using fieldsplit
+#if PETSC_VERSION_LESS_THAN(3,9,0)
+    PetscErrorCode __libmesh_petsc_DMCreateSubDM(DM dm, PetscInt numFields, PetscInt fields[], IS *is, DM *subdm)
```

Comment:
We've gone away from double-leading-underscore names for PETSc callback functions, see for example petsc_nonlinear_solver.h where they are now called `libmesh_petsc_snes_monitor`, `libmesh_petsc_snes_residual`, etc. The only reason for this is that all double-leading-underscore names are technically reserved for the compiler... there's really no danger of a name collision in this case, though.